File: WEBAPP/frontend/src/models/scenarios/scenarioViewModels.js
/**
 * @typedef {import('../simulation/simulationSostViewModel').SimulationSost} SimulationSost
 * @typedef {import('../simulation/simulationSpostaViewModel').SimulationSposta} SimulationSposta
 * @typedef {import('../weekly_programming/competitorProgramsViewModel').CompetitorProgramsViewModel} CompetitorProgramsViewModel
 */

/**
 * @typedef {Object} ScenarioViewModel
 * @property {string} id
 * @property {string} scenario_type
 * @property {string} program_id
 * @property {string} program_name
 * @property {string} program_channel
 * @property {string|null} program_date
 * @property {string|null} program_from_time
 * @property {string|null} program_to_time
 * @property {number|null} program_share_predict
 * @property {string|null} creation_date
 * @property {string|null} modified_date
 * @property {(SimulationSost|SimulationSposta)[]} simulations
 */

/**
 * @typedef {Object} ScenarioListViewModel
 * @property {ScenarioViewModel[]} scenarios
 * @property {number} total
 */

/**
 * @typedef {CompetitorProgramsViewModel} ScenCompetitorProgramsViewModel
 */

export {}

File: WEBAPP/api/app/view_models/scenarios/scenario_view_models.py
from __future__ import annotations

from dataclasses import dataclass, field

from app.models.scenario import Scenario
from app.models.simulation import SimulationSost
from app.view_models.simulation import SimulationSostViewModel, SimulationSpostViewModel


@dataclass
class ScenarioViewModel:
    
    id: str
    scenario_type: str           # 'sostituzione' | 'spostamento'
    program_id: str
    program_name: str
    program_channel: str
    program_date: str | None
    program_from_time: str | None
    program_to_time: str | None
    program_share_predict: float | None
    creation_date: str | None
    modified_date: str | None
    simulations: list[SimulationSostViewModel | SimulationSpostViewModel] = field(default_factory=list)

    @classmethod
    def MapScenarioViewModelFromScenario(cls, s: Scenario) -> "ScenarioViewModel":
        return cls(
            id=s.id,
            scenario_type=s.scenario_type,
            program_id=s.program_id,
            program_name=s.program_name,
            program_channel=s.program_channel,
            program_date=s.program_date,
            program_from_time=s.program_from_time,
            program_to_time=s.program_to_time,
            program_share_predict=s.program_share_predict,
            creation_date=s.creation_date,
            modified_date=s.modified_date,
            simulations=[
                SimulationSostViewModel.MapSimulationSostViewModelFromSimulationSost(sim)
                if isinstance(sim, SimulationSost)
                else SimulationSpostViewModel.MapSimulationSpostaViewModelFromSimulationSposta(sim)
                for sim in s.simulations
            ],
        )


@dataclass
class ScenarioListViewModel:
    scenarios: list[ScenarioViewModel] = field(default_factory=list)
    total: int = 0
